English - Basic Modules (#28)
haskellfoundation/HaskellSchool: en/lessons/basics/modules.md

> --- version: 1.4.1 title: Modules ---

Reviewer: ```suggestion
---
version: 1.0.0
title: Modules
---
```

We know from experience it's unruly to have all of our functions in the same
file and scope. In this lesson we're going to cover how to group entities in
modules, and group modules in components. At their most basic, modules are just
a namespace that allow us to aggregate common pieces of functionality and avoid
naming collisions. However, modules are use to build up more complex constructs
called components. Modules also offer a tool for encapsulation, by allowing us
to choose what we export.

{% include toc.html %}

## Entities, Modules, and Components

In order to understand modules we have to understand the elements that compose
a module, and what modules can be composed into. The elements that can be
exported from a module are called entities. In Haskell those are:

- Types and type synonyms
- Values: constants and functions
- Typeclasses

Modules are really just collections of these entities that are exported.
Modules can be combined through a graph of imports and exports. These

> dependency graphs can be packaged together as a component. The two most common

Reviewer: Would it be more appropriate to have "these dependency graphs" be singular?

components are libraries and executables. A library allows you to export your
modules for re-use in other codebases, and an executable consists of an
entrypoint (a Main module) to the dependency graph that can be run as a

> program. A good example of a library is `base` and `prelude`. These libraries

Reviewer: It's not clear here that `prelude` is a module exported by `base`. Changing this also likely requires changing the plurality of some other sentences.

ship with GHC. A good example of a third party library that requires a package

> manager to import is `lens`. A good example of executables are `xmonad` and

Reviewer: There are quite a few short sentences / instances of repetition here, which perhaps reduce cohesion. Something like ", whereas good examples of executables are `xmonad` and `pandoc`" may increase cohesion. 

Perhaps you could also vary the choice of the word "good", but not at the cost of accuracy.

`pandoc`. Haskell's most popular package managers also have executables that
allow them to be run from the command line: `cabal-install` and `stack`.

To reiterate, entities are our basic language-level building blocks. Modules

> are groupings of entities with some or all exported. Components are graphs of

Reviewer: This is extremely nit-picky, but the usage of two plurals here introduces ambiguity. It could either mean each component has one graph, or each component has multiple graphs.

Reviewer: The changes done in response to https://github.com/haskellfoundation/HaskellSchool/pull/28#discussion_r676190194 are enough to fix this.

modules that can be made available in other codebases or executed via an
entrypoint module.

## Module Syntax

Module syntax is driven by 2 keywords, `module` and `where`. The `module` keyword

> precedes the modules name, which must match the filename, and optionally a

Reviewer: ```suggestion
precedes the module's name, which must match the filename, and optionally a
```

collection of entities to export. The `where` keyword ends the module
declaration and indicates the beginning of the module.

Below is an example of a module declaration exposing all entities.

```haskell
-- src/Example.hs
module Example where

id :: a -> a
id a = a

compose :: (b -> c) -> (a -> b) -> (a -> c)
compose f g = \a -> f (g a)
```

Here is the same module, but residing at a different filepath.

```haskell
-- src/Data/Examples/Example.hs
module Data.Examples.Example where

id :: a -> a
id a = a

compose :: (b -> c) -> (a -> b) -> (a -> c)
compose f g = \a -> f (g a)
```

Modules can export a subset of the entities they contain. In the example below
only the type `Identity` (and its constructor) and `id` are exported, `compose`

> is completely isolated to the local module scope.

Reviewer: It's unfortunate, but for the flow of the lessons, I believe we're going to restrain from code examples with types and typeclasses here, given the lesson is right at the start of the tutorial. However, I think it would be beneficial to the reader for you to quickly mention them (e.g. in your listing of entities / as a note). 

The plan is to introduce how to import and export each entity (except for declarations) when we reach them.

Author: Really? I am not sure I agree with this decision. Perhaps typeclasses (which I believe are not included in any code examples), but I think seeing a type exported and imported will be more helpful to the user than the confusion that might arise. Type declarations are one of the first things one learns about in haskell.

Reviewer: Reconsidering it, I'd agree with you that your example has little potential to cause confusion, and comes with some benefits. However, a note (e.g. `__Note__: you'll learn about using and defining types in Haskell in [types](../types) and [algebraic data types](../algebraic-data-types).`) at the end of a codeblock may help the reader even more.

Author: Yeah, I agree, that's a good idea.

```haskell
-- src/Identity.hs
module Identity (
  Identity(..),
  id
) where

data Identity a = Identity a

id :: a -> a
id a = a

compose :: (b -> c) -> (a -> b) -> (a -> c)
compose f g = \a -> f (g a)
```

Modules can also export types, but hide constructors. This means (with respect
to the example below) that you can speak about things of type `Identity` but you

> cannot construct them yourself. These are sometimes referred to as opaque types.

Reviewer: Similarly, it's difficult to decide whether this should go in this lesson. It would definitely be helpful to mention opaque types at some point in Haskell School though.

```haskell
-- src/Identity.hs
module Identity (
  Identity,
  wrapIdentity
) where

data Identity a = Identity a

wrapIdentity :: a -> Identity a
wrapIdentity = Identity
```

There is only 1 piece of syntax that ever comes before module syntax, that is
language extensions. We won't go over what they do here, but just so you are

> familiar, this is what they look like.

Reviewer: There are a couple of comma splices here.

```haskell
{-# LANGUAGE OverloadedStrings #-}
module Identity (
  Identity,
  wrapIdentity
) where

data Identity a = Identity a

wrapIdentity :: a -> Identity a
wrapIdentity = Identity
```

## Import Syntax

Modules can be listed as a dependency within other modules. It is convention to
have these declarations at the top of a file, below the module declaration.

> This is an "open" import, it brings all entities exported by the module into

Reviewer: I think this is also a comma splice.

import Data.Maybe
```

Sometimes we want to be explicit about the entities that we are bringing into
scope. The syntax below brings in all declared entities, and hides all the
entities that are elided.

```haskell
module Example where

import Data.Maybe (fromMaybe)
```

Another technique is to qualify an import so that all use sites are tagged.

```haskell
module Example where

import qualified Data.Maybe

myFromMaybe = Data.Maybe.fromMaybe
```

Note, you can use qualifications even if a module is not imported as
`qualified`. This highlights that qualification enforces a namespace, but we can
always optionally declare the namespace. This is useful for resolving naming
collisions. In this example I will use ghci to demonstrate a name collision error.

```console?lang=haskell&prompt=ghci>,ghci|
ghci> import Data.Map
ghci> import Data.List
ghci> :t map
<interactive>:1:1: error:
    Ambiguous occurrence ‘map’
    It could refer to
       either ‘Data.Map.map’,
              imported from ‘Data.Map’
              (and originally defined in ‘Data.Map.Internal’)
           or ‘Prelude.map’,
              imported from ‘Prelude’ (and originally defined in ‘GHC.Base’)

> ghci> :t Data.Map.map

Reviewer: `ghci> :t Data.Map.map` is duplicated here. 

This is also a bit pedantic, but is it necessary to import `Data.List`? Similarly, could you change the line number of the error to `<interactive>:2:1:` (assuming `Data.List` is no longer imported)?

ghci> :t Data.Map.map
Data.Map.map :: (a -> b) -> Map k a -> Map k b
ghci> :t Data.List.map
Data.List.map :: (a -> b) -> [a] -> [b]
```

> If `Data.Maybe` is too long to type out, we can provide an alias for our import

Reviewer: Just to show idiomatic usage, would it perhaps be better if something that is commonly imported qualified, such as `Text`, were used?

--- version: 1.4.1 title: Modules ---

We know from experience it's unruly to have all of our functions in the same
file and scope. In this lesson we're going to cover how to group entities in
modules, and group modules in components. At their most basic, modules are just
a namespace that allow us to aggregate common pieces of functionality and avoid
naming collisions. However, modules are use to build up more complex constructs
called components. Modules also offer a tool for encapsulation, by allowing us
to choose what we export.

{% include toc.html %}

## Entities, Modules, and Components

In order to understand modules we have to understand the elements that compose
a module, and what modules can be composed into. The elements that can be
exported from a module are called entities. In Haskell those are:

- Types and type synonyms
- Values: constants and functions
- Typeclasses

Modules are really just collections of these entities that are exported.

> Modules can be combined through a graph of imports and exports, this is

Reviewer: Perhaps try to replace the comma before "this is": I think this is a comma splice.

sometimes called a dependency graph. These dependency graphs can be packaged
together as a component. The two most common components are libraries and
executables. A library allows you to export your modules for re-use in other
codebases, and an executable consists of an entrypoint (a Main module) to the
dependency graph that can be run as a program. A good example of a library is

> `base` and `prelude` these libraries ship with GHC, `lens` is a good example of

Reviewer: Punctuation is needed before "these libraries". Changing this would also require rewriting the comma before lens.

a third party library that requires a package manager to import. A good example
of executables are `xmonad` and `pandoc`. Haskell's most popular package
managers also have executables that allow them to be run from the command line;

> `cabal-install` and `stack`.

Reviewer: I would use a colon here, but I'd have to get a grammar book out to make sure I'm right 😛.

To reiterate, entities are our basic language-level building blocks. Modules
are groupings of entities with some or all exported. Components are graphs of
modules that can be made available in other codebases or executed via an
entrypoint module.

## Module Syntax

> Module syntax is driven by 2 keywords `module` and `where`. The `module` keyword

Reviewer: I think punctuation is needed before `module`.

precedes the modules name, which must match the filename, and optionally a
collection of entities to export. The `where` keyword ends the module
declaration and indicates the beginning of the module.

> Module declaration exposing all entities.

Reviewer: It may be unusual for the reader to be introduced to a snippet of code like this. A word like "here" may be useful here.

---
version: 1.0.0
title: Modules
---

We know from experience it's unruly to have all of our functions in the same
file and scope. In this lesson we're going to cover how to group entities in
modules, and group modules in components. At their most basic, modules are just

> a namespace that allow us to aggregate common pieces of functionality and avoid
> naming collisions. However, modules are use to build up more complex constructs

Reviewer: ```suggestion
a namespace that allows us to aggregate common pieces of functionality and avoid
naming collisions. However, modules are used to build up more complex constructs
```
It required three proof readings to spot this😅.

constructs called components. Modules also offer a tool for encapsulation, by
allowing us to choose what we export.

{% include toc.html %}

## Entities, Modules, and Components

In order to understand modules we have to understand the elements that compose
a module, and what modules can be composed into. The elements that can be
exported from a module are called entities. In Haskell those are:

- Types and type synonyms
- Values: constants and functions
- Typeclasses

Modules are really just collections of these entities that are exported.
Modules can be combined through a graph of imports and exports. This dependency
graph can be packaged as a component. The two most common components are
libraries and executables. A library allows you to export your modules for
re-use in other codebases, and an executable consists of an entrypoint (a Main
module) to the dependency graph that can be run as a program. A good example of

> a library is `base` which exports `prelude`. The `base` library ships with GHC,

Reviewer 2: ```suggestion
a library is `base` which exports the Prelude. The `base` library ships with GHC,
```

whereas a third party library that requires a package manager is `lens`. Well

> known haskell executables are `xmonad` and `pandoc`. Haskell's most popular

Reviewer 2: ```suggestion
known haskell executables are [`xmonad`](https://xmonad.org/), [`pandoc`](https://pandoc.org/) and [`shellcheck`](https://www.shellcheck.net). Haskell's most popular
```

To reiterate, an entity is our basic language-level building block. Modules
are groupings of entities with some or all exported. Components are graphs of
modules that can be made available in other codebases or executed via an
entrypoint module.

## Module Syntax

Module syntax is driven by 2 keywords, `module` and `where`. The `module` keyword
precedes the module's name, which must match the filename, and optionally a
collection of entities to export. The `where` keyword ends the module
declaration and indicates the beginning of the module.

Below is an example of a module declaration exposing all entities.

```haskell
-- src/Example.hs
module Example where

id :: a -> a
id a = a

compose :: (b -> c) -> (a -> b) -> (a -> c)
compose f g = \a -> f (g a)
```

Here is the same module, but residing at a different filepath.

```haskell
-- src/Data/Examples/Example.hs
module Data.Examples.Example where

id :: a -> a
id a = a

compose :: (b -> c) -> (a -> b) -> (a -> c)
compose f g = \a -> f (g a)
```

Modules can export a subset of the entities they contain. In the example below
only the type `Identity` (and its constructor) and `id` are exported, `compose`

> is completely isolated to the local module scope.

Reviewer: I think there's a comma splice here.

```haskell
-- src/Identity.hs
module Identity (
  Identity(..),
  id
) where

data Identity a = Identity a

id :: a -> a
id a = a

compose :: (b -> c) -> (a -> b) -> (a -> c)
compose f g = \a -> f (g a)
```
__Note__: you'll learn about using and defining types in Haskell in
[types](../types) and [algebraic data types](../algebraic-data-types).

Modules can also export types, but hide constructors. This means (with respect
to the example below) that you can speak about things of type `Identity` but you
cannot construct them yourself. These are sometimes referred to as opaque types.

```haskell
-- src/Identity.hs
module Identity (
  Identity,
  wrapIdentity
) where

data Identity a = Identity a

wrapIdentity :: a -> Identity a
wrapIdentity = Identity
```

> There is only one piece of syntax that ever comes before module syntax; that is

Reviewer 2: FIY: Actually, module documentation for Haddock also comes before module syntax. It's in the form:

```haskell
{-|
Module      : W
Description : Short description
Copyright   : (c) Some Guy, 2013
                  Someone Else, 2014
License     : GPL-3
Maintainer  : sample@email.com
Stability   : experimental
Portability : POSIX

Here is a longer description of this module, containing some
commentary with @some markup@.
-}
module W where
```

We can link to the haddock lesson when it's made (but that's not for basic track)

Author: I would argue that that isn't a piece of syntax, but added a another comment.

modules, and group modules in components. At their most basic, modules are just
a namespace that allows us to aggregate common pieces of functionality and
avoid naming collisions. However, modules are used to build up more complex
constructs called components. Modules also offer a tool for encapsulation, by
allowing us to choose what we export.

{% include toc.html %}

## Entities, Modules, and Components

In order to understand modules we have to understand the elements that compose
a module, and what modules can be composed into. The elements that can be
exported from a module are called entities. In Haskell those are:

- Types and type synonyms
- Values: constants and functions
- Typeclasses

Modules are really just collections of these entities that are exported.
Modules can be combined through a graph of imports and exports. This dependency
graph can be packaged as a component. The two most common components are
libraries and executables. A library allows you to export your modules for
re-use in other codebases, and an executable consists of an entrypoint (a Main
module) to the dependency graph that can be run as a program. A good example of

> a library is `base` which exports the Prelude. The `base` library ships with GHC,

Reviewer: ```suggestion
a library is `base`, which exports the Prelude. The `base` library ships with GHC,
```

Adding a comma here may increase clarity. Without a comma, "which exports the Prelude" forms a restrictive clause, which could confuse a reader, as they may believe a specific variant of `base` is being identified.

Reviewer: I've read an article that indicates that this is more of a problem in British English, but it still recommends using a comma.

only the type `Identity` (and its constructor) and `id` are exported. The
`compose` function is completely isolated to the local module scope.

```haskell
-- src/Identity.hs
module Identity (
  Identity(..),
  id
) where

data Identity a = Identity a

id :: a -> a
id a = a

compose :: (b -> c) -> (a -> b) -> (a -> c)
compose f g = \a -> f (g a)
```
__Note__: you'll learn about using and defining types in Haskell in
[types](../types) and [algebraic data types](../algebraic-data-types).

Modules can also export types, but hide constructors. This means (with respect
to the example below) that you can speak about things of type `Identity` but you
cannot construct them yourself. These are sometimes referred to as opaque types.

```haskell
-- src/Identity.hs
module Identity (
  Identity,
  wrapIdentity
) where

data Identity a = Identity a

wrapIdentity :: a -> Identity a
wrapIdentity = Identity
```

There is only one piece of syntax that ever comes before module syntax (aside
from a haddock module declaration comment, haddock is discussed in a separate
section); that is language extensions. We won't go over what they do here, but

> just so you are familiar this is what they look like.

Reviewer: I understand that the initial issue I opened involved specifying the full structure of a file, but I feel like going into the existence of Haddock may be a bit too clunky here. 

If the phrasing "only one piece of syntax" were removed, peculiarities (like Haddock and other pragmas) wouldn't need to be mentioned. 

I'm probably still in favour of mentioning language extensions, because they are used even in the basics page of the site, and it would be useful for the reader to understand how to enable them in other contexts.

Author: Agreed. I made changes removed the unnecessary bit about module syntax. I went in more depth about language extensions and overloaded strings in particular to try and assuage any confusion, I would be interested to hear if you think it adds more or less confusion.

```haskell
{-# LANGUAGE OverloadedStrings #-}
module Identity (
  Identity,
  wrapIdentity
) where

data Identity a = Identity a

wrapIdentity :: a -> Identity a
wrapIdentity = Identity
```

## Import Syntax

Modules can be listed as a dependency within other modules. It is convention to

> have these declarations at the top of a file, below the module declaration.

Reviewer: Isn't it necessary to put imports before top level declarations because of the parser?

(Sorry I forgot to put this in the review)

Author: Oh yeah, good call, bad wording on my part.

compose f g = \a -> f (g a)
```

Modules can export a subset of the entities they contain. In the example below
only the type `Identity` (and its constructor) and `id` are exported. The
`compose` function is completely isolated to the local module scope.

```haskell
-- src/Identity.hs
module Identity (
  Identity(..),
  id
) where

data Identity a = Identity a

id :: a -> a
id a = a

compose :: (b -> c) -> (a -> b) -> (a -> c)
compose f g = \a -> f (g a)
```
__Note__: you'll learn about using and defining types in Haskell in
[types](../types) and [algebraic data types](../algebraic-data-types).

Modules can also export types, but hide constructors. This means (with respect
to the example below) that you can speak about things of type `Identity` but you
cannot construct them yourself. These are sometimes referred to as opaque types.

```haskell
-- src/Identity.hs
module Identity (
  Identity,
  wrapIdentity
) where

data Identity a = Identity a

wrapIdentity :: a -> Identity a
wrapIdentity = Identity
```

There is an important piece of syntax, which comes before module syntax, and
that is language extensions. We won't exhaustively go over the functionality of

> individual extensions do here, but it is valuable to be able to recognize them.

Reviewer: ```suggestion
There is an important piece of syntax, which comes before module syntax, and
that is language extensions. We won't exhaustively go over the functionality of
individual extensions here, but it is valuable to be able to recognise them.
```
I think an extra "do" slipped in.

At a high level language extensions change and extend the behaviour of GHC. The
example below `OverloadedStrings` changes the behaviour of the built in string
syntax (double quotes) to be more general so that you can use different types /
representations of strings that behave in similar ways.

> > Note: If the way that Haskell deals with strings is confusing right now,

Reviewer: ```suggestion
> __Note__: If the way that Haskell deals with strings is confusing right now,
```

> don't worry. Pretty much every new haskeller goes through this. We choose
> to introduce it now because language extension syntax is important to
> recognize and OverloadedStrings is probably the most common language
> extension. String types and language extensions will be covered in more
> depth later!

```haskell
{-# LANGUAGE OverloadedStrings #-}
module Stringlike where

import Data.Text (Text)
import Data.ByteString (ByteString)

-- Everything works out because haskell is able
-- to infer what representation to use from the type!

-- Lists of characters
stringlikeString :: [Char]
stringlikeString = "string"

-- UTF8 encoded text
stringlikeUTF8 :: Text
stringlikeUTF8 = "string"

-- Packed binary representation
stringlikePackedBinary :: ByteString
stringlikePackedBinary = "string"
```

## Import Syntax

Modules can be listed as a dependency within other modules. You might not want
to think of imports in terms of the dependency chain they imply, in this case
an import can be though of as bringing some entities from the imported module

> into the scope of your current module. It is necessary to have these

Reviewer: The sentence beginning "You might not want to think" has a comma splice and is a bit clunky. It should be possible to remove a lot of words by using the word "rather".
